### DISCUSSION TITLE
What is the best way to handle business logic with the new Route Handlers?

### DISCUSSION THREAD

**Discussion by darwin403** [vercel/next.js#50378] [App Router]
We have custom middlewares similar to Express.js middlewares. However, with the changes to `middleware.ts` and `route.ts`. We no longer have a way to access `req` and `res` objects simulatenously that persists across the entire chain. 

I've also pointed a more detailed explanation here: https://github.com/vercel/next.js/discussions/50344

Look forward any assitance!

**Comment by devjiwonchoi** [comment]
Based on my understanding of your question, do `NextRequest` and `NextResponse` not fulfill your requirements?
It seems like you could handle requests and responses through [NextRequest](https://nextjs.org/docs/app/api-reference/functions/next-request) and [NextResponse](https://nextjs.org/docs/app/api-reference/functions/next-response), 
and I hope these could resolve your "We no longer have a way to access req and res objects simulatenously that persists across the entire chain."

**Reply by darwin403** [reply]
Not really, the `req,res` are not accesible simulatenously at each step of the middleware. Please look at the other discussion I've linked. It points out in detail.
